### DISCUSSION TITLE
How to run script with ObjectVars via API?

### DISCUSSION THREAD

**Discussion by llamafilm** [netbox-community/netbox#16255] [Q&A]
I'm trying to run a custom script with ObjectVars.  I pass the object ID numbers in the HTTP POST like this:
```
'{"data": {"site": 1, "location": 1, "device": 18, "delay": 0}, "commit": false}'
```

This doesn't work.  It seems like the script interprets these variables as integers instead of their object.  If I print the `data` variable to the log, it looks like this:
```
{'site': 1, 'location': 1, 'device': 18, 'delay': 0}
```

On the other hand, if I run the script from the webpage, it looks like this:

```
{'site': <Site: LA-LAI>, 'location': <Location: Holodeck Lounge>, 'device': <Device: Roku Ultra (18)>, 'delay': 0}
```

**Comment by bitcollector1** [comment]
I'm also interested in this; currently I pass string vars and turn them into objects once they are inside the script.

**Comment by mmfreitas** [comment]
I'm also interested in this, have you been able to figure if this is possible?

**Comment by karolyczovek** [comment]
At least you can call your script :)

I alway get 404 as return, no matter how i call it.

**Comment by GoepiGB** [comment]
I've mitigated the issue in the script like this:

`site = data['site']`
`if (isinstance(site, int)):`
`    site = Site.objects.get(id = site)`

So it works both ways via API and GUI :)

**Reply by RedLine89** [reply]
While this is going to work is this really how it is intended to work?

**Reply by GoepiGB** [reply]
I don't know; maybe it is intended so that netbox does not need to check the data for sanity in the API call.
If so, it should be mentioned in the docs so you can write your scripts accordingly, and you don't have to figure out why it works one way and not the other.

**Comment by rfdrake** [comment]
My take on this is that it's a bug, but one that not many people hit.  I haven't had many occasions to run a custom script via an API hook.  Now that I'm trying to do it, I'm hitting the same problem you guys are, only with MultiObjectVar rather than ObjectVar.

I've added the same workaround to my script.  If this isn't a bug, we need to get the documentation updated so this is easier to figure out.  I spent the first 5 hours head scratching this thinking my script was just wrong and surely there was some behind-the-scenes way to format the data so that netbox would accept it.

**Comment by GoepiGB** [comment]
It looks like this has been fixed with version 4.2.4
[#18480](https://github.com/netbox-community/netbox/issues/18480)

**Reply by llamafilm** [reply]
That only addresses the CLI, not the API.
